src/graphic/animation/spritesheet_animation.cc

```diff
@@ -0,0 +1,222 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ */
+
+#include "graphic/animation/spritesheet_animation.h"
+
+#include <cassert>
+#include <cstdio>
+#include <limits>
+#include <memory>
+
+#include <boost/algorithm/string/replace.hpp>
+#include <boost/format.hpp>
+
+#include "base/log.h"
+#include "base/macros.h"
+#include "graphic/graphic.h"
+#include "graphic/image.h"
+#include "graphic/playercolor.h"
+#include "graphic/texture.h"
+#include "io/filesystem/filesystem.h"
+#include "io/filesystem/layered_filesystem.h"
+#include "logic/game_data_error.h"
+#include "scripting/lua_table.h"
+
+/*
+==============================================================================
+
+SpriteSheetAnimation::MipMapEntry IMPLEMENTATION
+
+==============================================================================
+*/
+
+SpriteSheetAnimation::SpriteSheetMipMapEntry::SpriteSheetMipMapEntry(const std::string& file, int init_rows, int init_columns)
+   : Animation::MipMapEntry(), sheet(nullptr), playercolor_mask_sheet(nullptr), rows(init_rows), columns(init_columns), w(0), h(0), sheet_file(file), playercolor_mask_sheet_file("") {
+
+	assert(g_fs->file_exists(file));
+
+	playercolor_mask_sheet_file = file;
+	boost::replace_last(playercolor_mask_sheet_file, ".png", "_pc.png");
+	if (g_fs->file_exists(playercolor_mask_sheet_file)) {
+		has_playercolor_masks = true;
+	} else {
+		playercolor_mask_sheet_file = "";
+	}
+}
+
+
+void SpriteSheetAnimation::SpriteSheetMipMapEntry::ensure_graphics_are_loaded() const {
+	if (sheet == nullptr) {
+		const_cast<SpriteSheetMipMapEntry*>(this)->load_graphics();
+	}
+}
+
+void SpriteSheetAnimation::SpriteSheetMipMapEntry::load_graphics() {
+	sheet = g_gr->images().get(sheet_file);
+
+	if (!playercolor_mask_sheet_file.empty()) {
+		playercolor_mask_sheet = g_gr->images().get(playercolor_mask_sheet_file);
+
+		if (sheet->width() != playercolor_mask_sheet->width()) {
+			throw Widelands::GameDataError(
+			   "animation sprite sheet has width %d but playercolor mask sheet has width %d. The sheet's image is %s",
+			   sheet->width(), playercolor_mask_sheet->width(), sheet_file.c_str());
+		}
+		if (sheet->height() != playercolor_mask_sheet->height()) {
+			throw Widelands::GameDataError(
+			   "animation sprite sheet has height %d but playercolor mask sheet has height %d. The sheet's image is %s",
+			   sheet->height(), playercolor_mask_sheet->height(), sheet_file.c_str());
+		}
+	}
+
+	// Frame width and height
+	w = sheet->width() / columns;
+	h = sheet->height() / rows;
```

Comment:
Throw an error if (w × columns != sheet->width()) and same for the height

Reply:
Done.